zcash_primitives/src/sapling/address.rs

```diff
@@ -0,0 +1,118 @@
+use group::{Group, GroupEncoding};
+
+use super::{
+    keys::Diversifier,
+    note::{Note, Rseed},
+};
+
+/// A Sapling payment address.
+///
+/// # Invariants
+///
+/// `pk_d` is guaranteed to be prime-order (i.e. in the prime-order subgroup of Jubjub,
+/// and not the identity).
+#[derive(Clone, Debug)]
+pub struct PaymentAddress {
+    pk_d: jubjub::SubgroupPoint,
+    diversifier: Diversifier,
+}
+
+impl PartialEq for PaymentAddress {
+    fn eq(&self, other: &Self) -> bool {
+        self.pk_d == other.pk_d && self.diversifier == other.diversifier
+    }
+}
+
+impl Eq for PaymentAddress {}
+
+impl PaymentAddress {
+    /// Constructs a PaymentAddress from a diversifier and a Jubjub point.
+    ///
+    /// Returns None if `pk_d` is the identity.
+    pub fn from_parts(diversifier: Diversifier, pk_d: jubjub::SubgroupPoint) -> Option<Self> {
```

Comment:
Pre-existing issue: this does not check that `diversifier` is valid, i.e. it does not check that `diversifier.g_d()` succeeds (which it may not since `Diversifier` is only a `pub` wrapper for `[u8; 11]`).

Reply:
Indeed. Fixing this is a change to the public API (both here, and in places where we no longer need to fallibly derive `g_d`), so I'm not doing it in this PR but plan to do so.